cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java

```diff
@@ -155,30 +156,32 @@ private int addBrokerMetricSamples(Cluster cluster,
         LOG.warn("Skip generating broker metric sample for broker {} because all broker metrics are missing.", node.id());
         continue;
       } else if (!brokerLoad.allBrokerMetricsAvailable()) {
-        if (brokerLoad.missingBrokerMetrics().size() == 0) {
+        if (brokerLoad.missingBrokerMetricsInMinSupportedVersion().size() == 0) {
           LOG.warn("Skip generating broker metric sample for broker {} because there are not enough topic metrics to "
                   + "generate broker metrics.", node.id());
         } else {
-          LOG.warn("Skip generating broker metric sample for broker {} because the following metrics are missing {}.",
-              node.id(), brokerLoad.missingBrokerMetrics());
+          LOG.warn("Skip generating broker metric sample for broker {} because the following required metrics are missing {}.",
+              node.id(), brokerLoad.missingBrokerMetricsInMinSupportedVersion());
         }
         continue;
       }
 
       boolean validSample = true;
-      BrokerMetricSample brokerMetricSample = new BrokerMetricSample(node.host(), node.id());
-      for (RawMetricType rawBrokerMetricType : RawMetricType.brokerMetricTypes()) {
-        // We require the broker to report all the metric types. Otherwise we skip the broker.
-        if (!brokerLoad.brokerMetricAvailable(rawBrokerMetricType)) {
-          skippedBroker++;
-          validSample = false;
-          LOG.warn("Skip generating broker metric sample for broker {} because it does not have %s metrics or "
-                        + "the metrics are inconsistent.", node.id(), rawBrokerMetricType);
-          break;
-        } else {
-          MetricInfo metricInfo = brokerMetricDef.metricInfo(KafkaMetricDef.forRawMetricType(rawBrokerMetricType).name());
-          double metricValue = brokerLoad.brokerMetric(rawBrokerMetricType);
-          brokerMetricSample.record(metricInfo, metricValue);
+      BrokerMetricSample brokerMetricSample = new BrokerMetricSample(node.host(), node.id(), brokerLoad._brokerSampleDeserializationVersion);
+      for (Map.Entry<Byte, List<RawMetricType>> entry : RawMetricType.brokerMetricTypesDiffByVersion().entrySet()) {
```

Comment:
I am wondering whether this check is necessary. I `brokerLoad.allBrokerMetricsAvailable()` is enough to generate a broker metric sample. Won't this check make the V4 reporter's metric always get skipped?

Reply:
The `BrokerLoad#setNullableBrokerMetrics()` will ensure that `v4`'s `nullable` metrics will be populated. Hence, we will not skip the raw metrics. This sanity check is different from the `brokerLoad.allBrokerMetricsAvailable()` in the sense that the former checks for the existence of minimum required broker metrics availability while the latter is about ensuring that each metric either has a value or has correctly been nullified.

I think we can make this more clear by changing `allBrokerMetricsAvailable()`  to `minRequiredBrokerMetricsAvailable()`